src/http.cc

```diff
@@ -959,6 +959,8 @@ HttpStateData::haveParsedReplyHeaders()
 
         case ReuseDecision::cachePositively:
             entry->makePublic();
+            if (!entry->preparePublicEntry())
```

Comment:
I agree that we may need to create a Transients entry in this context.

However, the combination of making the entry public, then checking whether we actually can make it public, followed by undoing that making (when the check fails) looks weird do me. Why not check _before_ we make the entry public?!

Furthermore, if all makePublic() calls by writers require the same check-make logic, then this logic should be encapsulated in a new method. Please see whether we should split makePublic() into makePublicForWriting() and makePublicForReading() methods so that all makePublic() callers are forced to make the right call, without the risk of forgetting to follow up with the preparePublicEntry() and makePrivate() calls.

Reply:
> Why not check before we make the entry public?!

This 'checking' consists of Transients::startWriting() call and its result, which expects already existing public StoreEntry as a parameter.

Reply:
> Please see whether we should split makePublic() into makePublicForWriting() and makePublicForReading()

The 'make public for reading' phrase looks strange to me: we do not 'make public' entries, just loaded entry from a cache because they are public already. In this case, we just assign existing public keys with them (calling hashInsert()).

Reply:
> Transients::startWriting() [expects an] existing public StoreEntry as a parameter.

AFAICT, we can either compute the public key inside Transients::startWriting() OR supply the computed public key to that method (whichever results in overall better code). There may be other, better solutions as well, of course.

The low-level pieces we have to build with remain about the same, but, to ensure the new "writers have Transients entry" invariant, we have to rearrange/add higher level methods so that the callers do not need to remember (the correct order of) all the necessary events that constitute preparing an entry for writing.

> The 'make public for reading' phrase looks strange to me: we do not 'make public' entries, just loaded entry from a cache because they are public already. In this case, we just assign existing public keys with them (calling hashInsert()).

I probably agree. I have no emotional attachment to makePublicForWriting() and makePublicForReading() names. We may be able to find better name(s) when the methods are implemented and used. As always, some other adjustments may be needed as well. The key here is that multiple high-level callers should not be forced to remember the correct sequence of checks/calls to make their entries public. Those callers should call some function or method that will orchestrate the right sequence of checks/calls for them.